### PR TITLE
[SPARK-33003][PYTHON][DOCS] Add type hints guidelines to the documentation

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

Add type hints guidelines to developer docs.

### Why are the changes needed?

Since it is a new and still somewhat evolving feature, we should provided clear guidelines for potential contributors.

### Does this PR introduce _any_ user-facing change?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Looks pretty good. Let me know when you think it's ready @zero323

File: python/docs/source/development/contributing.rst [hunk 1]
```diff
@@ -77,6 +77,43 @@ There are a couple of additional notes to keep in mind when contributing to code
 * Be Pythonic.
 * APIs are matched with Scala and Java sides in general.
 * PySpark specific APIs can still be considered as long as they are Pythonic and do not conflict with other existent APIs, for example, decorator usage of UDFs.
+* If you extend or modify public API, please adjust corresponding type hints. See `Contributing and Maintaining Type Hints`_ for details.
+
+Contributing and Maintaining Type Hints
+----------------------------------------
+
+PySpark type hints are provided using stub files, placed in the same directory as the annotated module, with exception to ``# type: ignore`` in modules which don't have their own stubs (tests, examples, non-public API).
```
Comment (HyukjinKwon) [inline python/docs/source/development/contributing.rst:85]: just a nit but maybe: `tests, examples, non-public API` -> `tests, examples and non-public API`?

**Comment by zero323** [comment]
Thanks for checking this @HyukjinKwon! I think it is ready for review ‒ I've modified grammar as requested and section about validating.

As of user part, I decided against adding it for now. I have some hints, but these seem to be too localized (specific editors, common typing issues), while for most of the users, benefits will be mostly transparent. Maybe we could add a short guide for different environments (I have [this answer](https://stackoverflow.com/a/34714207) on Stack, which seems to be weirdly popular) and add some typing notes there?

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by zero323** [comment]
Thanks @HyukjinKwon!

**Pull request closed without merging**